lib/azure/armrest/configuration.rb

```diff
@@ -108,23 +110,26 @@ def initialize(args)
 
         options.each { |key, value| send("#{key}=", value) }
 
-        unless client_id && client_key && tenant_id && subscription_id
-          raise ArgumentError, "client_id, client_key, tenant_id and subscription_id must all be specified"
+        unless client_id && client_key && tenant_id
+          raise ArgumentError, "client_id, client_key, and tenant_id must all be specified"
         end
 
         # Allows for URI objects or Strings.
         @proxy = @proxy.to_s if @proxy
```

Comment:
@djberg96 we need to re-visit all `attr_accessor` attributes as some are inappropriate. Do we allow them all to be modified after the configuration object is constructed? If not, we should make them read-only and allow them to be set only through constructor. For example, if subscription_id is set after initialization, then you won't get the chance to validate_subscription and fetch providers.

If we do allow them to be changed at anytime, I would change at least `@proxy` and `@subscription_id` to be `attr_reader` and provide customized setters, where you can force `@proxy` to be string and validate subscription and fetch providers.